src/TransformationUtils.cc

```diff
@@ -88,57 +88,63 @@ std::unique_ptr<TransitionSystem> toTransitionSystem(ChcDirectedGraph const & gr
     return ts;
 }
 
-bool strongConnection(std::unordered_set<int> & visitedVertices, std::unordered_set<int> & verticesOnStack,
+// This function follows a Tarjan's strong connection detection algorithm:
+// https://en.wikipedia.org/wiki/Tarjan's_strongly_connected_components_algorithm
+// If some vertice is added on stack and it was already visited, then the strong connection is detected
+// It was updated to remember the nodes which participate in the loop (lines 123, 127)
+void strongConnection(std::unordered_set<SymRef, SymRefHash> & visitedVertices,
+                      std::unordered_set<SymRef, SymRefHash> & verticesOnStack,
                       AdjacencyListsGraphRepresentation const & graphRepresentation, ChcDirectedGraph const & graph,
-                      SymRef node) {
-    visitedVertices.insert(node.x);
-    verticesOnStack.insert(node.x);
+                      SymRef node, std::vector<EId> & loop) {
+    visitedVertices.insert(node);
+    verticesOnStack.insert(node);
     auto const & outEdges = graphRepresentation.getOutgoingEdgesFor(node);
     if (size(outEdges) <= 1) {
-        verticesOnStack.erase(node.x);
-        return false;
+        verticesOnStack.erase(node);
+        return;
```

Comment:
Why can you ignore nodes with one outgoing edge?

Reply:
Because we simplify loop beforehand, so if there is just one out-edge it is selfloop, defacto this condition can be changed to `size(outEdges) == 0` and it would work the same